release-25.2: tree: add a session variable to get pre-25.2 variadic function typing

mw5h opened the pull request:
Backport 1/1 commits from #144522.

/cc @cockroachdb/release

---

In 25.2 we're changing the handling of variadic builtin functions to be more like Postgres. Instead of the AnyElement behavior that CRDB has had prior to this release, variadic builtins will accept variable elements of type Any. Any differs from AnyElement in that Any is truly unconstrained, where all AnyElement parameters to a function must have the same type.

While the Any behavior is more flexible than AnyElement, it also means that the type checker can't infer argument types in some circumstances, which could lead to existing workloads to stop working properly. This patch introduces the 'use_pre_25_2_variadic_builtins' session variable and the 'sql.defaults.use_pre_25_2_variadic_builtins' cluster setting to get the old behavior back.

Fixes: #143359

Release justification: This is a low risk safety valve to turn off a feature already in 25.2.

mw5h committed:
tree: add a session variable to get pre-25.2 variadic function typing

In 25.2 we're changing the handling of variadic builtin functions to be
more like Postgres. Instead of the AnyElement behavior that CRDB has had
prior to this release, variadic builtins will accept variable elements
of type Any. Any differs from AnyElement in that Any is truly
unconstrained, where all AnyElement parameters to a function must have
the same type.

While the Any behavior is more flexible than AnyElement, it also means
that the type checker can't infer argument types in some circumstances,
which could lead to existing workloads to stop working properly. This
patch introduces the 'use_pre_25_2_variadic_builtins' session variable
and the 'sql.defaults.use_pre_25_2_variadic_builtins' cluster setting to
get the old behavior back.

Fixes: #143359
Release note (sql change): Release 25.2 updates the behavior of
polymorphic builtin functions to be more like Postgres. These
functions include: concat, concat_ws, format, json_build_array,
json_build_objects, jsonb_build_array, jsonb_build_objects,
num_notnulls, num_nulls, pg_column_size.  Specifically, the variable
arguments to these functions no longer have to all have the same
type. As a result, CockroachDB's type inference engine will in some
cases no longer be able to infer argument types where it previously
could. While the new behavior matches Postgres, there is a possibility
that previously working CockroachDB applications will encounter new
errors. This patch introduces a new session variable
'use_pre_25_2_variadic_builtins' to restore the old behavior (and
limitations).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144600)
<!-- Reviewable:end -->

mw5h committed:
tree: fix a test that wasn't seeing the error that it should have

An astute reviewer noted that the builtin_function logic test had a
PREPARE with no EXECUTE and asked if that was intentional. It turns out
that it was, but the PREPARE was expected to fail!

After some investigation, it was discovered that resetting
use_pre_25_2_variadic_builtins was not causing CONCAT to return to the
normal 25.2 behavior, which would be to error. After some poking, I
discovered that the second PREPARE was not being planned at all because
it was hitting the query plan cache. Adding the session variable to the
function that tests memo staleness addressed the issue.

Informs: #143359
Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for doing this!

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144600#-:-OO1DciOCwPYfzALczIC:b-pt8m11)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144600#-OO1CIsjCLWsojOw9kki:-OO1CIsjCLWsojOw9kkj:bf577ud):*
Does the `sql.defaults...` cluster setting still exist? Or maybe the commit message is out of date?
___
*[`-- commits` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144600#-OO1DfnH6Hy4daqfi3RT:-OO1DfnH6Hy4daqfi3RU:bohenpm):*
Looking at #141193 I think the full list is:

* concat
* concat_ws
* crdb_internal.datums_to_bytes
* format
* json_build_array
* json_build_objects
* jsonb_build_array
* jsonb_build_objects
* num_notnulls
* num_nulls
* pg_column_size

It would probably be good to list these all explicitly in the release note.
___
*[`pkg/sql/sem/tree/type_check.go` line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144600#-OO1Fm2x0e57kVopsR5z:-OO1Fm2x0e57kVopsR6-:bu2rgf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/56edcd900e85fc19b4e4388f331c6df196364ab8/pkg/sql/sem/tree/type_check.go#L78)):*
> ```Go
> 	UnsupportedTypeChecker UnsupportedTypeChecker
> 
> 	// UsePre25_2VariadicBuiltins is set to true when we should use the pre-25.2
> ```

nit: comment doesn't match name


<!-- Sent from Reviewable.io -->

rytaft commented:
Don't forget to add #144603 to this backport before merging
